include/cetl/unbounded_variant.hpp

```diff
@@ -556,27 +557,55 @@ struct base_access : base_storage<Pmr, Footprint, Alignment>
     template <typename Tp, std::enable_if_t<is_rtti_convertible<Tp>, int> = 0>
     void make_converters() noexcept
     {
-        value_const_converter_ = [](const void* const storage, const type_id& id) {
-            const auto ptr = static_cast<const Tp*>(storage);
-            return ptr->_cast_(id);
+        value_const_converter_ = [](const void* const storage, const cetl::type_id& dst_type_id) {
+            CETL_DEBUG_ASSERT(nullptr != storage, "");
+            const auto        ptr     = static_cast<const Tp*>(storage);
+            const void* const dst_ptr = ptr->_cast_(dst_type_id);
+            return std::make_pair(dst_ptr, cetl::type_id_value<Tp>);
         };
-        value_converter_ = [](void* const storage, const type_id& id) {
-            auto ptr = static_cast<Tp*>(storage);
-            return ptr->_cast_(id);
+        value_mut_converter_ = [](void* const storage, const cetl::type_id& dst_type_id) {
+            CETL_DEBUG_ASSERT(nullptr != storage, "");
+            const auto ptr = static_cast<Tp*>(storage);
+            return ptr->_cast_(dst_type_id);
         };
     }
 
     template <typename Tp, std::enable_if_t<!is_rtti_convertible<Tp>, int> = 0>
     void make_converters() noexcept
     {
-        value_const_converter_ = [](const void* const storage, const type_id& id) {
-            return (id == type_id_value<Tp>) ? storage : nullptr;
+        value_const_converter_ = [](const void* const storage, const cetl::type_id& dst_type_id) {
+            CETL_DEBUG_ASSERT(nullptr != storage, "");
+            const void* const dst_ptr = (dst_type_id == cetl::type_id_value<Tp>) ? storage : nullptr;
+            return std::make_pair(dst_ptr, cetl::type_id_value<Tp>);
         };
-        value_converter_ = [](void* const storage, const type_id& id) {
-            return (id == type_id_value<Tp>) ? storage : nullptr;
+        value_mut_converter_ = [](void* const storage, const cetl::type_id& dst_type_id) {
+            CETL_DEBUG_ASSERT(nullptr != storage, "");
+            return (dst_type_id == cetl::type_id_value<Tp>) ? storage : nullptr;
         };
     }
 
+    /// \brief Returns the unique identifier of the actual type of the stored value.
+    ///        `cetl::type_id_invalid` if storage is empty.
+    ///
+    CETL_NODISCARD cetl::type_id type_id() const noexcept
+    {
+        if (!has_value())
+        {
+            return cetl::type_id_invalid;
+        }
+        CETL_DEBUG_ASSERT(nullptr != value_const_converter_, "Non-empty storage is expected to have value converter.");
+
+        return value_const_converter_(base::get_raw_storage(), {}).second;
+    }
+
+    /// \brief Returns the size of the stored value in bytes.
+    ///        Zero if storage is empty.
+    ///
+    CETL_NODISCARD std::size_t type_size() const noexcept
+    {
+        return has_value() ? base::get_value_size() : 0UL;
+    }
```

Comment:
Both these methods are very useful at least at libcyphal unit tests. But both of them could be useful also at production code - to have a peek what value (or size) is currently stored at ub_var.

Reply:
Also, note that there is no size penalty for these two - size is already these, and querying of type id is done by reusing existing const convertor.